refactor: better README in bigquery table access pattern analysis (#679)
GoogleCloudPlatform/professional-services: examples/bigquery-table-access-pattern-analysis/README.md

<i>Note that each of the source table(s) can be a different table type, it can be either a view, materialized view, or a normal table. We only consider the normal tables because the notion of 'analysing the update pattern' does not really apply to view and materialized view. View does not gets updated, and materialized view gets updated automatically</i>

We can categorise every pipeline to different pipeline type and scheduling patern. There are 3 different pipeline types, namely:
<ul>
<li>Live pipelines: Pipelines that run regularly on an obvious schedule, until now
<li>Dead pipelines: Pipelines that used to run regularly on an obvious schedule, but it stopped some time ago
<li>Ad hoc pipelines: Pipelines that does not have a regular scheduling pattern detected, or not enough repetition to conclude that it’s a scheduled live pipeline
</ul>

The scheduling pattern can be identified as hourly, daily, weekly, or non-deterministically (no obvious pattern).

#### <b>Purpose</b>

> This tool helps identify pipeline optimisation points, by pinpointing tables with high difference of writing and reading frequency throughout the data warehouse queries. High discrepancy between the write and read frequency of a table might be a good starting point for identifying optimisation points. Using this tool, we will also be able to visualise the pipelines that are involved with a table of interest, you can then further analyse the pipeline type and its scheduling pattern, and also each of the job of each pipelines, and pinpoint problems or optimisations from there. 

Reviewer: ```suggestion
This tool helps identify tables with a high difference between the write and read frequency across data warehouse queries. High discrepancy between the write and read frequency of a table might be a good starting point for identifying optimisation points. Using this tool we can visualise the pipelines that are involved with a table of interest. We can then further analyse the pipeline type, its scheduling pattern and the jobs in each pipeline, and pinpoint problems or optimisations from there. 
```

![](assets/pipeline-example.gif)

#### <b>Analysing the Result</b>
As can be seen from the GIF, the tool will visualise all the pipelines associated with a table. To be specific, this includes all query jobs that has this table of interest as its source or destination table. As mentioned above, for every query job, there are source table(s) and also a single destination table. 

The tables that are involved in the pipelines associated with a table (all tables that are included when the pipeline graph of a table of interest is plotted) employs the below logic:

For every query jobs that has the table of interest as one of its source tables or destination table, 
* For every source table(s) of every query job that has the table of interest as one of its source table(s), recursively find query jobs that has this source table as its destination table, and get its source table(s). 
* For every destination table of every query job that has the table of interest as its destination table, recursively find query jobs that has destination table as its source table, and get its destination table.

> As seen from the GIF too, for every tables that are involved in the pipeline of the table of interest, you can toggle to it, and see the details of the job schedule of every query involving this particular table. It will list down all pipelines, the ad-hoc, live or dead pipelines that has this table as its source table, and destination table. 

Reviewer: ```suggestion
As seen from the GIF, for every table that is involved in the pipeline of the table of interest, you can select it, and see the details of the job schedule of every query involving this particular table. It will list all ad-hoc, live or dead pipelines that have this table as its source or destination table. 
```

        "name": "timestamp",
        "type": "TIMESTAMP",
        "mode": "NULLABLE",

>         "description": "Timestamp on when the job was run"

Reviewer: ```suggestion
        "description": "Timestamp when the job was run"
```

        "name": "timestamps",
        "type": "ARRAY<TIMESTAMP>",
        "mode": "NULLABLE",

>         "description": "Timestamps on when this pipeline was run in the past"

Reviewer: ```suggestion
        "description": "Timestamps when this pipeline was run in the past"
```

        "name": "pipelineId",
        "type": "INTEGER",
        "mode": "NULLABLE",

>         "description": "The pipeline ID of pipeline that this pair was part of"

Reviewer: ```suggestion
        "description": "The pipeline ID for the pipeline that this pair was part of"
```

<li>getPipelineTypeAndSchedule

> This funtion takes in an array of timestamp, and return a struct of the pipeline type and schedule according to the history. There are 3 possible values for pipeline type: live/dead/ad hoc, and there are 4 possible values for schedule: non deterministic/hourly/daily/monthly. 

Reviewer: ```suggestion
This function takes an array of timestamps and returns a struct of the pipeline type and schedule according to the history. There are 3 possible values for pipeline type: live/dead/ad hoc, and there are 4 possible values for schedule: non deterministic/hourly/daily/monthly. 
```
